### ISSUE TITLE
onRegionDidChange triggers too many times

### ISSUE THREAD

**Issue by savv** [nitaliano/react-native-mapbox-gl#1108]
If I drag the map with multiple short drags, and I start dragging the map before it has "lost its momentum", it onRegionDidChange seems to trigger for each initiated drag.

(Because I store the center of the map in redux and because I then use that state to set centerCoordinate this results in the map jumping around.)

**Comment by Gp2mv3** [comment]
I confirm that the event is sent multiple time as explained in my previous issue #1098 (in Android at least).

**Comment by kristfal** [comment]
We kind of have a way to solve OPs issue. Every onRegionDidChange event has a flag indicating if it is a user initiated event. 

In the case of your redux store, you can listen to move events as before, but add a check in your map wrapper component to only set centerCoordinate when user interaction is false or when you set it programmatically via your redux actions. 

That way you don’t get into the recursive loop you’re seeing now. 

As for the event firing multiple times, this is something that should be fixed, but it’s quite tricky as the events propagate from the iOS and Android native side, and those differ quite a bit at the moment.

**Comment by savv** [comment]
Thank you. I haven't observed a recursive loop yet (it might be there, but I haven't observed a centerCoordinates change triggering onRegionDidChange). I have only seen two queued events triggering in rapid succession - which as you say should not happen. onRegionDidChange should only fire for the last one.

**Comment by nitaliano** [comment]
Yes, it is something that should be fixed, we are at the mercy off the native sdks here so we can try to update our throttle code to try to filter out these extra events better, but the best solution will be having it fixed upstream in the Android SDK

**Comment by savv** [comment]
To be clear, I've been having this issue in iOS.

**Comment by StasDoskalenko** [comment]
I've been using debounce functions to resolve this issue. As a bonus, we've got some optimization.

**Comment by nitaliano** [comment]
I added support for debouncing in this PR https://github.com/mapbox/react-native-mapbox-gl/pull/1132 you can specify the time in ms that you want to debounce the event for `onRegionWillChange` and `onRegionDidChange`

regionwillchange defaults to 0
regiondidchange defaults to 500

**Comment by nitaliano** [comment]
Merged the PR into master

**Comment by nbolten** [comment]
This issue is still present even with debouncing - debouncing just hides it for some amount of cases.

I'm running into the same issues listed by the issue poster and the comment on the merged pull request: a map "flick" (map pans with "momentum") fires two onregiondidchange events: one part-way through and one when the map actually stops.

This makes it difficult to control the map center programmatically, as too many state changes happen and the map can even get out of sync with state due to some internal map debouncing.

I attempted to work around this issue by filtering the events, trying to only keep the second one (they could be distinguished by the "animated" property), and this worked fine until I needed to catch zoom changes using the same event, and the "animated" property situation is reversed in that case.

I ended up creating a "last event" debouncer that prevents the redux action from firing until after map moving events have stopped for some time. This still has issues with jumpiness and still gets out of sync with state, so this is still an issue that impacts my projects.

I'd be happy with more information on the map move events to make this issue easier to work around. Having accurate and consistent "animated" properties would help.

**Comment by chrusart** [comment]
@nbolten  check PR #1310
